README.md
>2019.07.14    
>Currently, I have graduated from campus and doing another kind of job. So this project may not be updated again.

## Face_Pytorch
The implementation of  popular face recognition algorithms in pytorch framework, including arcface, cosface and sphereface and so on.

All codes are evaluated on Pytorch 0.4.0 with Python 3.6, Ubuntu 16.04.10, CUDA 9.1 and CUDNN 7.1. Partially evaluated on Pytorch 1.0.	

## Implementation & Modifications
* re-implemented by [Ardihikaru]
* Environment Setup
    - Install Anaconda3
    - Create conda env: `$ conda create -n facepy python=3.6 -c conda-forge` 
    - Create conda env: `$ conda create -n facepy python=3.7 -c conda-forge` 
    - Activate Conda Env: `$ conda activate facepy`
    - Install all dependencies
      1. Install MXNet: `$ pip install mxnet`
      2. Install Pillow: `$ pip install Pillow`
      3. Install OpenCV: `$ pip install opencv-python`
      4. Install [TQDM](https://github.com/tqdm/tqdm): `$ pip install tqdm`
      5. Install PyTorch: `$ conda install -c pytorch pytorch`
      6. Install PytorchVision: `$ conda install -c pytorch torchvision`
      7. Install visdom: `$ pip install visdom`
    - Download sample Dataset.
      1. Create folder in `$ROOT/data/`
      2. Download [dataset](https://drive.google.com/file/d/1npu8Ma9BZDp4Z18ARkitwP0OozAy__aG/view) and copy only `part0_train.bin` into `$ROOT/data/`
    - Extract `.bin` file from [Learn-to-cluster](https://github.com/ardihikaru/learn-to-cluster):
      `$ python utils/load_images_from_bin.py --source-data bin --dataset-path data/part0_train.bin --output-path data/output_bin/`

python3 -m pip install numpy -I

## Data Preparation
For CNN training, I use CASIA-WebFace and Cleaned MS-Celeb-1M, aligned by MTCNN with the size of 112x112.  For performance testing, I report the results on LFW, AgeDB-30, CFP-FP, MegaFace rank1 identification and verification.

For AgeDB-30 and CFP-FP, the aligned images and evaluation images pairs are restored from the mxnet binary file provided by [insightface](https://github.com/deepinsight/insightface), tools are available in this repository. You should install a mxnet-cpu first for the image parsing, just do ' **pip install mxnet** ' is ok.  
[LFW @ BaiduNetdisk](https://pan.baidu.com/s/1Rue4FBmGvdGMPkyy2ZqcdQ),   [AgeDB-30 @ BaiduNetdisk](https://pan.baidu.com/s/1sdw1lO5JfP6Ja99O7zprUg),   [CFP_FP @ BaiduNetdisk](https://pan.baidu.com/s/1gyFAAy427weUd2G-ozMgEg)

## Results
> MobileFaceNet: Struture described in MobileFaceNet  
> ResNet50: Original resnet structure  
> ResNet50-IR: CNN described in ArcFace paper  
> SEResNet50-IR: CNN described in ArcFace paper 
### Verification results on LFW, AgeDB-30 and CFP_FP  
Small Protocol: trained with CASIA-WebFace of data size: 453580/10575  
Large Protocol: trained with DeepGlint MS-Celeb-1M of data size: 3923399/86876

  Model Type    |   Loss    |   LFW   | AgeDB-30 | CFP-FP | Model Size | protocol
:--------------:|:---------:|:-------:|:--------:|:-------|:----------:|:--------:
MobileFaceNet   |  ArcFace  |  99.23  |   93.26  |  94.34 |    4MB     |  small
ResNet50-IR     |  ArcFace  |  99.42  |   94.45  |  95.34 |   170MB    |  small  
SEResNet50-IR   |  ArcFace  |  99.43  |   94.50  |  95.43 |   171MB    |  small
MobileFaceNet   |  ArcFace  |  99.58  |   96.57  |  92.90 |    4MB     |  large
ResNet50-IR     |  ArcFace  |  99.82  |   98.07  |  95.34 |   170MB    |  large
SEResNet50-IR   |  ArcFace  |  99.80  |   98.13  |  95.60 |   171MB    |  large
ResNet100-IR    |  ArcFace  |  99.83  |   98.28  |  96.41 |   256MB    |  large 

There exists an odd result fact that when training under small protocol, CFP-FP performances better than AgeDB-30, while when training with large scale dataset, CFP-FP performances worse than AgeDB-30.

### MegaFace rank 1 identifiaction accuracy and verfication@FPR=1e-6 results

  Model Type    |   Loss    | MF Acc. | MF Ver. | MF Acc.@R | MF Ver.@R | SIZE  | protocol
:--------------:|:---------:|:-------:|:-------:|:---------:|:---------:|:-----:|:-------:
MobileFaceNet   |  ArcFace  |  69.10  |  84.23  |   81.15   |   85.86   |  4MB  |  small
ResNet50-IR     |  ArcFace  |  74.31  |  88.23  |   87.44   |   89.56   | 170MB |  small
SEResNet50-IR   |  ArcFace  |  74.37  |  88.32  |   88.30   |   89.65   | 171MB |  small
MobileFaceNet   |  ArcFace  |  74.95  |  88.77  |   89.47   |   91.03   |  4MB  |  large
ResNet50-IR     |  ArcFace  |  79.61  |  96.02  |   96.58   |   96.78   | 170MB |  large
SEResNet50-IR   |  ArcFace  |  79.91  |  96.10  |   97.01   |   97.60   | 171MB |  large
ResNet100-IR    |  ArcFace  |  80.40  |  96.94  |   97.60   |   98.05   | 256MB |  large



## Usage 
1. Download the source code to your machine.
2. Prepare the train dataset and train list, test dataset and test verification pairs.
3. Set your own dataset path and any other parameters in train.py.
4. Run train.py file, test accuracy will print into log file during training process. 
---
5. Every evaluate file can work independently for the model test process. just set your own args in the file. 

## Visualization
Visdom support for loss and accuracy during training process.  
![avatar](result/visualization.jpg)

 
Softmax Loss vs Softmax_Center Loss. Left: softmax training set. Right: softmax + center loss training set. 
<div align="center">
  <img src="result/softmax.gif" alt="train" width="49.5%">
  <img src="result/softmax_center.gif" alt="train" width="49.5%">
</div>

## References
[MuggleWang/CosFace_pytorch](https://github.com/MuggleWang/CosFace_pytorch)  
[Xiaoccer/MobileFaceNet_Pytorch](https://github.com/Xiaoccer/MobileFaceNet_Pytorch)  
[TreB1eN/InsightFace_Pytorch](https://github.com/TreB1eN/InsightFace_Pytorch)  
[deepinsight/insightface](https://github.com/deepinsight/insightface)  
[KaiyangZhou/pytorch-center-loss](https://github.com/KaiyangZhou/pytorch-center-loss)  
[tengshaofeng/ResidualAttentionNetwork-pytorch](https://github.com/tengshaofeng/ResidualAttentionNetwork-pytorch)

## Todo
1. Report the test results on DeepGlint Trillion Pairs Challenge.
2. Add C++ api for fast deployment with pytorch 1.0.
3. Train the ResNet100-based model.


CHANGELOG.md
## 0.1.0 (十二月 26, 2019)
  - added bump-version.sh and modified .gitignore files
  - Update README.md
  - update readme
  - update readme
  - readme
  - update
  - train.py
  - multi margin
  - train
  - readme
  - update
  - update
  - update
  - readme
  - eval_deepglint
  - eval lfw
  - train
  - update
  - cbam
  - cbam
  - update
  - update
  - train.py
  - cbam
  - update
  - self attention
  - self attention
  - self attention
  - self_attention
  - eval megaface
  - shit code
  - inner product
  - eval
  - update
  - eval megaface
  - timer to calculate the inference time
  - code optimization
  - fix a bug
  - update
  - update
  - update
  - sram
  - lfw loader
  - restore image from mxnet file
  - restore the train data from mxnet data type
  - readme
  - readme
  - cpp api
  - cppapi
  - test
  - self attention
  - update readme
  - attention 56/92
  - update for pytorch 1.0.1
  - update
  - train
  - cbam
  - eval
  - lfw_blufr_test
  - senet
  - cbam and senet
  - readme
  - visualization
  - visualization
  - readme
  - visualization
  - train cbam
  - center loss
  - CBAM
  - channel attention and spatial attention
  - train
  - train SeRes100-IR
  - visualization
  - eval
  - train res50-ir
  - eval
  - update initialization for center loss
  - add visualization support during train
  - add bn for spherenet
  - visualization
  - plot loss and accuracy during training
  - plot loss and accuracy during training
  - visdom support for visualization
  - update
  - add train accuracy
  - original softmax training
  - center loss
  - center loss
  - center loss
  - inner product
  - inner product
  - center loss
  - train
  - theta distribution
  - plot theta distribution
  - theta distribution
  - save net head
  - readme
  - result
  - result
  - training log
  - test SEResNet50-IR
  - result
  - pretrain
  - train
  - result
  - readme
  - result
  - result
  - train accuracy
  - update
  - train
  - eval
  - result
  - readme
  - train accuracy
  - eval
  - result
  - eval code
  - arcfacenet
  - result
  - sphereface margin
  - result
  - sphereface net
  - multiplicative angular margin
  - train
  - readme
  - evaluate the test datasets
  - test
  - result
  - result
  - result
  - result
  - result
  - add Res_IR_SE backbone
  - result
  - cfp-fp evaluation
  - update
  - result
  - result
  - result
  - load image from mxnet file
  - result
  - result
  - result
  - result
  - restore image from mxnet binary
  - agedb30
  - train
  - update the lfw evaluate methods
  - result
  - update the project architecture
  - update dataset
  - train
  - line coding
  - fxxk
  - all
  - encoding
  - readme
  - readme
  - backbone
  - megaface feature extractor
  - lfw
  - train
  - resnet for backboke
  - readme
  - readme
  - train on gpu server
  - cosface margin
  - cosine margin product
  - cuda
  - gpu init
  - readme
  - train
  - train mobilefacenet
  - random flip for traindata
  - train
  - lfw
  - logging tool
  - casia webface
  - update
  - train
  - a
  - readme
  - test
  - mobilefacenet
  - package init
  - pycharm
  - Initial commit



utils/load_images_from_bin.py
#!/usr/bin/env python
# encoding: utf-8
'''
@author: wujiyang
@contact: wujiyang@hust.edu.cn
@file: load_images_from_bin.py
@time: 2018/12/25 19:21
@desc: For AgeDB-30 and CFP-FP test dataset, we use the mxnet binary file provided by insightface, this is the tool to restore
       the aligned images from mxnet binary file.
       You should install a mxnet-cpu first, just do 'pip install mxnet==1.2.1' is ok.
'''

# from PIL import Image
import cv2
import os
import pickle
import mxnet as mx
from tqdm import tqdm
import argparse

'''
For train dataset, insightface provide a mxnet .rec file, just install a mxnet-cpu for extract images
'''

def load_mx_rec(rec_path):
    save_path = os.path.join(rec_path, 'emore_images_2')
    if not os.path.exists(save_path):
        os.makedirs(save_path)

    imgrec = mx.recordio.MXIndexedRecordIO(os.path.join(rec_path, 'train.idx'), os.path.join(rec_path, 'train.rec'), 'r')
    img_info = imgrec.read_idx(0)
    header,_ = mx.recordio.unpack(img_info)
    max_idx = int(header.label[0])
    for idx in tqdm(range(1,max_idx)):
        img_info = imgrec.read_idx(idx)
        header, img = mx.recordio.unpack_img(img_info)
        label = int(header.label)
        #img = cv2.cvtColor(img, cv2.COLOR_RGB2BGR)
        #img = Image.fromarray(img)
        label_path = os.path.join(save_path, str(label).zfill(6))
        if not os.path.exists(label_path):
            os.makedirs(label_path)
        #img.save(os.path.join(label_path, str(idx).zfill(8) + '.jpg'), quality=95)
        cv2.imwrite(os.path.join(label_path, str(idx).zfill(8) + '.jpg'), img)


def load_image_from_bin(bin_path, save_dir):
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)
    file = open(os.path.join(save_dir, '../', 'lfw_pair.txt'), 'w')
    bins, issame_list = pickle.load(open(bin_path, 'rb'), encoding='bytes')
    for idx in tqdm(range(len(bins))):
        _bin = bins[idx]
        img = mx.image.imdecode(_bin).asnumpy()
        img = cv2.cvtColor(img, cv2.COLOR_RGB2BGR)
        cv2.imwrite(os.path.join(save_dir, str(idx+1).zfill(5)+'.jpg'), img)
        if idx % 2 == 0:
            label = 1 if issame_list[idx//2] == True else -1
            file.write(str(idx+1).zfill(5) + '.jpg' + ' ' + str(idx+2).zfill(5) +'.jpg' + ' ' + str(label) + '\n')


if __name__ == '__main__':
    # parser = argparse.ArgumentParser(description='Testing')
    # parser.add_argument('--root', type=str, default='/media/sda/lfw/lfw_align_112', help='The path of lfw data')
    # parser.add_argument('--file_list', type=str, default='/media/sda/lfw/pairs.txt', help='The path of lfw data')
    # parser.add_argument('--backbone_net', type=str, default='CBAM_100_SE', help='MobileFace, CBAM_50, CBAM_50_SE, CBAM_100, CBAM_100_SE')
    # parser.add_argument('--feature_dim', type=int, default=512, help='feature dimension')
    # parser.add_argument('--resume', type=str, default='./model/SERES100_SERES100_IR_20190528_132635/Iter_342000_net.ckpt',
    #                     help='The path pf save model')
    # parser.add_argument('--feature_save_path', type=str, default='./result/cur_epoch_lfw_result.mat',
    #                     help='The path of the extract features save, must be .mat file')
    # parser.add_argument('--gpus', type=str, default='1,3', help='gpu list')
    # args = parser.parse_args()

    #bin_path = 'D:/face_data_emore/faces_webface_112x112/lfw.bin'
    save_dir = 'data/output_bin'
    # rec_path = 'D:/face_data_emore/faces_emore'
    # rec_path = 'data/lfw.bin'
    bin_path = 'data/lfw.bin'
    # load_mx_rec(rec_path)
    load_image_from_bin(bin_path, save_dir)


train_v2.py
#!/usr/bin/env python
# encoding: utf-8
'''
@author: wujiyang
@contact: wujiyang@hust.edu.cn
@file: train.py.py
@time: 2018/12/21 17:37
@desc: train script for deep face recognition
'''

import os
import torch.utils.data
from torch.nn import DataParallel
from datetime import datetime
from backbone.mobilefacenet import MobileFaceNet
from backbone.cbam import CBAMResNet
from backbone.attention import ResidualAttentionNet_56, ResidualAttentionNet_92
from margin.ArcMarginProduct import ArcMarginProduct
from margin.MultiMarginProduct import MultiMarginProduct
from margin.CosineMarginProduct import CosineMarginProduct
from margin.InnerProduct import InnerProduct
from utils.visualize import Visualizer
from utils.logging import init_log
from dataset.casia_webface import CASIAWebFace
from dataset.lfw import LFW
from dataset.agedb import AgeDB30
from dataset.cfp import CFP_FP
from torch.optim import lr_scheduler
import torch.optim as optim
import time
from eval_lfw import evaluation_10_fold, getFeatureFromTorch
import numpy as np
import torchvision.transforms as transforms
import argparse


def train(args):
    # gpu init
    multi_gpus = False
    if len(args.gpus.split(',')) > 1:
        multi_gpus = True
    os.environ['CUDA_VISIBLE_DEVICES'] = args.gpus
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    # log init
    save_dir = os.path.join(args.save_dir, args.model_pre + args.backbone.upper() + '_' + datetime.now().strftime('%Y%m%d_%H%M%S'))
    if os.path.exists(save_dir):
        raise NameError('model dir exists!')
    os.makedirs(save_dir)
    logging = init_log(save_dir)
    _print = logging.info

    # dataset loader
    transform = transforms.Compose([
        transforms.ToTensor(),  # range [0, 255] -> [0.0,1.0]
        transforms.Normalize(mean=(0.5, 0.5, 0.5), std=(0.5, 0.5, 0.5))  # range [0.0, 1.0] -> [-1.0,1.0]
    ])
    # validation dataset
    trainset = CASIAWebFace(args.train_root, args.train_file_list, transform=transform)
    trainloader = torch.utils.data.DataLoader(trainset, batch_size=args.batch_size,
                                              shuffle=True, num_workers=8, drop_last=False)
    # test dataset
    lfwdataset = LFW(args.lfw_test_root, args.lfw_file_list, transform=transform)
    lfwloader = torch.utils.data.DataLoader(lfwdataset, batch_size=128,
                                             shuffle=False, num_workers=4, drop_last=False)
    # agedbdataset = AgeDB30(args.agedb_test_root, args.agedb_file_list, transform=transform)
    # agedbloader = torch.utils.data.DataLoader(agedbdataset, batch_size=128,
    #                                         shuffle=False, num_workers=4, drop_last=False)
    # cfpfpdataset = CFP_FP(args.cfpfp_test_root, args.cfpfp_file_list, transform=transform)
    # cfpfploader = torch.utils.data.DataLoader(cfpfpdataset, batch_size=128,
    #                                           shuffle=False, num_workers=4, drop_last=False)

    # define backbone and margin layer
    if args.backbone == 'MobileFace':
        net = MobileFaceNet()
    elif args.backbone == 'Res50_IR':
        net = CBAMResNet(50, feature_dim=args.feature_dim, mode='ir')
    elif args.backbone == 'SERes50_IR':
        net = CBAMResNet(50, feature_dim=args.feature_dim, mode='ir_se')
    elif args.backbone == 'Res100_IR':
        net = CBAMResNet(100, feature_dim=args.feature_dim, mode='ir')
    elif args.backbone == 'SERes100_IR':
        net = CBAMResNet(100, feature_dim=args.feature_dim, mode='ir_se')
    elif args.backbone == 'Attention_56':
        net = ResidualAttentionNet_56(feature_dim=args.feature_dim)
    elif args.backbone == 'Attention_92':
        net = ResidualAttentionNet_92(feature_dim=args.feature_dim)
    else:
        print(args.backbone, ' is not available!')

    if args.margin_type == 'ArcFace':
        margin = ArcMarginProduct(args.feature_dim, trainset.class_nums, s=args.scale_size)
    elif args.margin_type == 'MultiMargin':
        margin = MultiMarginProduct(args.feature_dim, trainset.class_nums, s=args.scale_size)
    elif args.margin_type == 'CosFace':
        margin = CosineMarginProduct(args.feature_dim, trainset.class_nums, s=args.scale_size)
    elif args.margin_type == 'Softmax':
        margin = InnerProduct(args.feature_dim, trainset.class_nums)
    elif args.margin_type == 'SphereFace':
        pass
    else:
        print(args.margin_type, 'is not available!')

    if args.resume:
        print('resume the model parameters from: ', args.net_path, args.margin_path)
        net.load_state_dict(torch.load(args.net_path)['net_state_dict'])
        margin.load_state_dict(torch.load(args.margin_path)['net_state_dict'])

    # define optimizers for different layer
    criterion = torch.nn.CrossEntropyLoss().to(device)
    optimizer_ft = optim.SGD([
        {'params': net.parameters(), 'weight_decay': 5e-4},
        {'params': margin.parameters(), 'weight_decay': 5e-4}
    ], lr=0.1, momentum=0.9, nesterov=True)
    exp_lr_scheduler = lr_scheduler.MultiStepLR(optimizer_ft, milestones=[6, 11, 16], gamma=0.1)

    if multi_gpus:
        net = DataParallel(net).to(device)
        margin = DataParallel(margin).to(device)
    else:
        net = net.to(device)
        margin = margin.to(device)


    best_lfw_acc = 0.0
    best_lfw_iters = 0
    # best_agedb30_acc = 0.0
    # best_agedb30_iters = 0
    # best_cfp_fp_acc = 0.0
    # best_cfp_fp_iters = 0
    total_iters = 0
    vis = Visualizer(env=args.model_pre + args.backbone)
    for epoch in range(1, args.total_epoch + 1):
        exp_lr_scheduler.step()
        # train model
        _print('Train Epoch: {}/{} ...'.format(epoch, args.total_epoch))
        net.train()

        since = time.time()
        for data in trainloader:
            img, label = data[0].to(device), data[1].to(device)
            optimizer_ft.zero_grad()

            raw_logits = net(img)
            output = margin(raw_logits, label)
            total_loss = criterion(output, label)
            total_loss.backward()
            optimizer_ft.step()

            total_iters += 1
            # print train information
            if total_iters % 100 == 0:
                # current training accuracy
                _, predict = torch.max(output.data, 1)
                total = label.size(0)
                correct = (np.array(predict.cpu()) == np.array(label.data.cpu())).sum()
                time_cur = (time.time() - since) / 100
                since = time.time()
                vis.plot_curves({'softmax loss': total_loss.item()}, iters=total_iters, title='train loss',
                                xlabel='iters', ylabel='train loss')
                vis.plot_curves({'train accuracy': correct / total}, iters=total_iters, title='train accuracy', xlabel='iters',
                                ylabel='train accuracy')

                _print("Iters: {:0>6d}/[{:0>2d}], loss: {:.4f}, train_accuracy: {:.4f}, time: {:.2f} s/iter, learning rate: {}".format(total_iters, epoch, total_loss.item(), correct/total, time_cur, exp_lr_scheduler.get_lr()[0]))

            # save model
            if total_iters % args.save_freq == 0:
                msg = 'Saving checkpoint: {}'.format(total_iters)
                _print(msg)
                if multi_gpus:
                    net_state_dict = net.module.state_dict()
                    margin_state_dict = margin.module.state_dict()
                else:
                    net_state_dict = net.state_dict()
                    margin_state_dict = margin.state_dict()
                if not os.path.exists(save_dir):
                    os.mkdir(save_dir)
                torch.save({
                    'iters': total_iters,
                    'net_state_dict': net_state_dict},
                    os.path.join(save_dir, 'Iter_%06d_net.ckpt' % total_iters))
                torch.save({
                    'iters': total_iters,
                    'net_state_dict': margin_state_dict},
                    os.path.join(save_dir, 'Iter_%06d_margin.ckpt' % total_iters))

            # test accuracy
            if total_iters % args.test_freq == 0:

                # test model on lfw
                net.eval()
                getFeatureFromTorch('./result/cur_lfw_result.mat', net, device, lfwdataset, lfwloader)
                lfw_accs = evaluation_10_fold('./result/cur_lfw_result.mat')
                _print('LFW Ave Accuracy: {:.4f}'.format(np.mean(lfw_accs) * 100))
                if best_lfw_acc <= np.mean(lfw_accs) * 100:
                    best_lfw_acc = np.mean(lfw_accs) * 100
                    best_lfw_iters = total_iters

                # test model on AgeDB30
                # getFeatureFromTorch('./result/cur_agedb30_result.mat', net, device, agedbdataset, agedbloader)
                # age_accs = evaluation_10_fold('./result/cur_agedb30_result.mat')
                # _print('AgeDB-30 Ave Accuracy: {:.4f}'.format(np.mean(age_accs) * 100))
                # if best_agedb30_acc <= np.mean(age_accs) * 100:
                #     best_agedb30_acc = np.mean(age_accs) * 100
                #     best_agedb30_iters = total_iters

                # test model on CFP-FP
                # getFeatureFromTorch('./result/cur_cfpfp_result.mat', net, device, cfpfpdataset, cfpfploader)
                # cfp_accs = evaluation_10_fold('./result/cur_cfpfp_result.mat')
                # _print('CFP-FP Ave Accuracy: {:.4f}'.format(np.mean(cfp_accs) * 100))
                # if best_cfp_fp_acc <= np.mean(cfp_accs) * 100:
                #     best_cfp_fp_acc = np.mean(cfp_accs) * 100
                #     best_cfp_fp_iters = total_iters
                # _print('Current Best Accuracy: LFW: {:.4f} in iters: {}, AgeDB-30: {:.4f} in iters: {} and CFP-FP: {:.4f} in iters: {}'.format(
                #     best_lfw_acc, best_lfw_iters, best_agedb30_acc, best_agedb30_iters, best_cfp_fp_acc, best_cfp_fp_iters))

                # vis.plot_curves({'lfw': np.mean(lfw_accs), 'agedb-30': np.mean(age_accs), 'cfp-fp': np.mean(cfp_accs)}, iters=total_iters,
                #                 title='test accuracy', xlabel='iters', ylabel='test accuracy')

                vis.plot_curves({'lfw': np.mean(lfw_accs)}, iters=total_iters,
                                title='test accuracy', xlabel='iters', ylabel='test accuracy')
                net.train()

    # _print('Finally Best Accuracy: LFW: {:.4f} in iters: {}, AgeDB-30: {:.4f} in iters: {} and CFP-FP: {:.4f} in iters: {}'.format(
    #     best_lfw_acc, best_lfw_iters, best_agedb30_acc, best_agedb30_iters, best_cfp_fp_acc, best_cfp_fp_iters))
    _print('Finally Best Accuracy: LFW: {:.4f} in iters: {}'.format(
        best_lfw_acc, best_lfw_iters))
    print('finishing training')


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='PyTorch for deep face recognition')
    parser.add_argument('--train_root', type=str, default='data/faces112x112', help='train image root')
    parser.add_argument('--train_file_list', type=str, default='data/train_val/train.txt', help='train list')
    parser.add_argument('--lfw_test_root', type=str, default='data/faces112x112', help='lfw image root')
    parser.add_argument('--lfw_file_list', type=str, default='data/faces112x112/pairs.txt', help='lfw pair file list')
    # parser.add_argument('--agedb_test_root', type=str, default='/media/sda/AgeDB-30/agedb30_align_112', help='agedb image root')
    # parser.add_argument('--agedb_file_list', type=str, default='/media/sda/AgeDB-30/agedb_30_pair.txt', help='agedb pair file list')
    # parser.add_argument('--cfpfp_test_root', type=str, default='/media/sda/CFP-FP/cfp_fp_aligned_112', help='agedb image root')
    # parser.add_argument('--cfpfp_file_list', type=str, default='/media/sda/CFP-FP/cfp_fp_pair.txt', help='agedb pair file list')

    parser.add_argument('--backbone', type=str, default='SERes100_IR', help='MobileFace, Res50_IR, SERes50_IR, Res100_IR, SERes100_IR, Attention_56, Attention_92')
    parser.add_argument('--margin_type', type=str, default='ArcFace', help='ArcFace, CosFace, SphereFace, MultiMargin, Softmax')
    parser.add_argument('--feature_dim', type=int, default=512, help='feature dimension, 128 or 512')
    parser.add_argument('--scale_size', type=float, default=32.0, help='scale size')
    # parser.add_argument('--batch_size', type=int, default=200, help='batch size')
    parser.add_argument('--batch_size', type=int, default=32, help='batch size')
    # parser.add_argument('--total_epoch', type=int, default=18, help='total epochs')
    parser.add_argument('--total_epoch', type=int, default=10, help='total epochs')

    parser.add_argument('--save_freq', type=int, default=3000, help='save frequency')
    parser.add_argument('--test_freq', type=int, default=3000, help='test frequency')
    parser.add_argument('--resume', type=int, default=False, help='resume model')
    parser.add_argument('--net_path', type=str, default='', help='resume model')
    parser.add_argument('--margin_path', type=str, default='', help='resume model')
    parser.add_argument('--save_dir', type=str, default='./model', help='model save dir')
    parser.add_argument('--model_pre', type=str, default='SERES100_', help='model prefix')
    # parser.add_argument('--gpus', type=str, default='0,1,2,3', help='model prefix')
    parser.add_argument('--gpus', type=str, default='0', help='model prefix')

    args = parser.parse_args()

    train(args)


